File: Problems/Two Pointers/3Sum Closest/solution.py
from typing import List
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        closest_diff = 500001
        for i in range(n-2):
            st, end = i+1, n-1
            while st < end:
                triplet_sum = nums[st] + nums[end] + nums[i]
                if triplet_sum == target:
                    return target
                elif triplet_sum  < target:
                    st = st + 1
                else:
                    end = end - 1
                if abs(target-triplet_sum) < abs(closest_diff):
                    closest_diff = target-triplet_sum
        return target - closest_diff

File: Problems/Two Pointers/3Sum Closest/solution.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        int closest_diff = 500001;
        for(int i = 0; i < n-2; i++){
            int st = i+1, end = n-1;
            while(st < end){
                int triplet_sum = nums[i] + nums[st] + nums[end];
                if(triplet_sum == target) return target;
                else if(triplet_sum < target) st++;
                else end--;
                if(abs(target - triplet_sum) < abs(closest_diff)) closest_diff = target-triplet_sum;
            }
        }
        return target-closest_diff;
    }
};